torch/_inductor/ir.py

```diff
@@ -6686,6 +6689,202 @@ def codegen_collective(self, wrapper, output_name, input_names):
         )
 
 
+# TODO(yifu): replace the CollectiveKernel IR hierarchy with _CollectiveKernel.
+class _CollectiveKernel(FallbackKernel):
+    @classmethod
+    def make_fallback_kernel(cls, layout, kernel, inputs, *args, **kwargs):
+        with V.graph.fake_mode:
+            (
+                example_outputs,
+                tensor_args,
+                non_tensor_args,
+                unflatten_args,
+                schema,
+            ) = cls.process_kernel(kernel, inputs, *args, **kwargs)
+        return (
+            cls(
+                layout,
+                kernel,
+                tensor_args,
+                non_tensor_args,
+                unflatten_args,
+                schema=schema,
+            ),
+            example_outputs,
+        )
+
+    @staticmethod
+    def wrap_inplace_input(inp):
+        return TensorBox.create(
+            InPlaceHint(
```

Comment:
This is a separate discussion, but it was never clear to me the purpose of `InplaceHint`? Why is it needed?

Reply:
> This is a separate discussion, but it was never clear to me the purpose of InplaceHint?

@wconstab gave a nice explanation below. More concretely, during codegen, if the wrapper decides to reuse the input buffer for `InPlaceHint`, the latter effectively aliases the former. The reuse decision conveniently skip the allocation. Otherwise, a buffer is allocated for `InPlaceHint`, and the data of the input buffer is copied to the newly allocated buffer.

> Why is it needed?

This was to achieve oppurtunistic de-functionalization at IR level. It works, but I agree post grad passes are more suitable for this.